File: Top_Interview_Questions/121._Best_Time_to_Buy_and_Sell_Stock/solution.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # we can keep track of the minimum price and the maximum profit we've seen so far
        # this way, we ensure that the purchase comes before the selling
        min_price = float('inf')
        max_profit = 0
        
        for price in prices:
            if price < min_price:
                min_price = price
            else:
                profit = price - min_price
                if profit > max_profit:
                    max_profit = profit
        
        return max_profit

File: Top_Interview_Questions/121._Best_Time_to_Buy_and_Sell_Stock/solution.java
class Solution {
    public int maxProfit(int[] prices) {
        // hold the value of the minimum price
        // and update the max profit
        int minPrice = Integer.MAX_VALUE;
        int mp = 0;

        for (int price : prices) {
            if (price < minPrice)
                minPrice = price;
            else
                mp = Math.max(mp, price - minPrice);
        }

        return mp;
    }
}